### PR TITLE
implemented getaddrinfo

### PR DESCRIPTION
**Pull request by Kolpa**
### First getaddrinfo implementation.
* AI_CANONNAME ist the only working flag right now.
* port as String and int should be working as intended
* family and type arguments are ignored in this implementation

### CODE REVIEW

File: graalpython/com.oracle.graal.python/src/com/oracle/graal/python/builtins/modules/SocketModuleBuiltins.java [hunk 1]
```diff
@@ -40,31 +40,128 @@
...
+import java.io.BufferedReader;
+import java.io.File;
+import java.io.FileNotFoundException;
+import java.io.FileReader;
+import java.net.Inet4Address;
+import java.net.Inet6Address;
 import java.net.InetAddress;
 import java.net.UnknownHostException;
-import java.util.List;
+import java.util.*;
```
Comment (timfel) [inline graalpython/com.oracle.graal.python/src/com/oracle/graal/python/builtins/modules/SocketModuleBuiltins.java]: No wildcards imports please

File: graalpython/com.oracle.graal.python/src/com/oracle/graal/python/builtins/modules/SocketModuleBuiltins.java [hunk 2]
```diff
@@ -113,4 +210,79 @@ String doGeneric() {
             }
         }
     }
+
+    @Builtin(name = "getaddrinfo", parameterNames = {"host", "port", "family", "type", "proto", "flags"})
+    @GenerateNodeFactory
+    public abstract static class GetAddrInfoNode extends PythonBuiltinNode {
+        @Specialization
+        @TruffleBoundary
```
Comment (timfel) [inline graalpython/com.oracle.graal.python/src/com/oracle/graal/python/builtins/modules/SocketModuleBuiltins.java]: TruffleBoundary is not needed here.

File: graalpython/com.oracle.graal.python/src/com/oracle/graal/python/builtins/modules/SocketModuleBuiltins.java [hunk 3]
```diff
@@ -113,4 +210,79 @@ String doGeneric() {
...
+        Object getAddrInfo(String host, int port, int family, int type, int proto, PInt flags) {
+            return getAddrInfo(host, port, family, type, proto, flags.asInt());
+        }
+        @Specialization
+        @TruffleBoundary
```
Comment (timfel) [inline graalpython/com.oracle.graal.python/src/com/oracle/graal/python/builtins/modules/SocketModuleBuiltins.java]: TruffleBoundary is not needed here.

File: graalpython/com.oracle.graal.python/src/com/oracle/graal/python/builtins/modules/SocketModuleBuiltins.java [hunk 4]
```diff
@@ -113,4 +210,79 @@ String doGeneric() {
...
+        Object getAddrInfo(String host, String port, int family, int type, int proto, PInt flags) {
+            return getAddrInfo(host, port, family, type, proto, flags.asInt());
+        }
+        @Specialization
+        @TruffleBoundary
+        Object getAddrInfo(String host, int port, int family, int type, int proto, int flags) {
+            InetAddress[] adresses = resolveHost(host);
+            List<Service> serviceList = new ArrayList();
```
Comment (timfel) [inline graalpython/com.oracle.graal.python/src/com/oracle/graal/python/builtins/modules/SocketModuleBuiltins.java]: It looks like you're not modifying the serviceList in `mergeAddressesAndServices`. Can you instead use
```
Service[] serviceList = new ServiceList[2];
serviceList[0] = new Service(port, "tcp");
serviceList[1] = new Service(port, "udp");
```
?

Comment (timfel) [reply graalpython/com.oracle.graal.python/src/com/oracle/graal/python/builtins/modules/SocketModuleBuiltins.java]: If you can, the TruffleBoundary here can also be removed

File: graalpython/com.oracle.graal.python/src/com/oracle/graal/python/builtins/modules/SocketModuleBuiltins.java [hunk 5]
```diff
@@ -113,4 +210,79 @@ String doGeneric() {
...
+            if (services == null) {
+                services = parseServices();
+            }
+            List<Service> serviceList = services.get(port);
+
+            InetAddress[] adresses = resolveHost(host);
+            return mergeAdressesAndServices(adresses, serviceList, proto, flags);
+        }
+
+        private Object mergeAdressesAndServices(InetAddress[] adresses, List<Service> serviceList, int proto, int flags) {
```
Comment (timfel) [inline graalpython/com.oracle.graal.python/src/com/oracle/graal/python/builtins/modules/SocketModuleBuiltins.java:247]: This needs a TruffleBoundary (uses ArrayList). My general rule is to move the TruffleBoundary to the most specific method possible, so all the callers don't need it

**Review by timfel — COMMENTED** [review]
In general this looks good, the issues are basically nitpicks :)

File: graalpython/com.oracle.graal.python/src/com/oracle/graal/python/builtins/modules/SocketModuleBuiltins.java (hunk 4 above)
Comment (milanpro) [reply graalpython/com.oracle.graal.python/src/com/oracle/graal/python/builtins/modules/SocketModuleBuiltins.java]: > It looks like you're not modifying the serviceList in `mergeAddressesAndServices`. Can you instead use
> 
> ```
> Service[] serviceList = new ServiceList[2];
> serviceList[0] = new Service(port, "tcp");
> serviceList[1] = new Service(port, "udp");
> ```
> 
> ?

Problem is that parseServices needs the ArrayLists add, so to change this one, we would need to convert those lists in the services Map too.

**Pull request closed without merging**